rust-src/concordium_base/src/protocol_level_tokens/token_event.rs

```diff
@@ -1,40 +1,80 @@
-use super::{cbor::RawCbor, TokenId};
+use crate::transactions::Memo;
 
-/// Event produced from the effect of a token holder transaction.
+use super::{cbor::RawCbor, TokenAmount, TokenHolder, TokenId};
+
+/// Details provided by the token module in the event of rejecting a
+/// transaction.
 #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct TokenHolderEvent {
+pub struct TokenModuleRejectReason {
     /// The unique symbol of the token, which produced this event.
     pub token_id:   TokenId,
     /// The type of event produced.
     pub event_type: TokenEventType,
     /// The details of the event produced, in the raw byte encoded form.
-    pub details:    RawCbor,
+    pub details:    Option<RawCbor>,
 }
 
-/// Event produced from the effect of a token governance transaction.
+/// An event produced from the effect of a token transaction.
 #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct TokenGovernanceEvent {
+pub struct TokenEvent {
     /// The unique symbol of the token, which produced this event.
-    pub token_id:   TokenId,
+    pub token_id: TokenId,
+    /// The type of the event.
+    pub event:    Event,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+/// The type of the token event.
+pub enum Event {
+    /// An event emitted by the token module.
+    Module(TokenModuleEvent),
+    /// An event emitted when a transfer of tokens is performed.
+    Transfer(TokenTransferEvent),
+    /// An event emitted when the token supply is updated by minting tokens to a
+    /// token holder.
+    Mint(TokenSupplyUpdateEvent),
+    /// An event emitted when the token supply is updated by burning tokens from
+    /// the balance of a token holder.
+    Burn(TokenSupplyUpdateEvent),
+}
+
+/// Event produced from the effect of a token holder transaction.
```

Comment:
```suggestion
/// Event produced from the effect of a token transaction.
```

E.g. updating the allow/deny list will emit events. Meaning `token governance transations` can emit this event as well.